### ISSUE TITLE
Nonsense warnings about using `next` to skip iteration

### ISSUE THREAD

**Issue by YorickPeterse** [bbatsov/rubocop#1238]
Consider the following snippet of Ruby code:

```ruby
foos.each do
  unless foo
    yield bar
  end
end
```

When running Rubocop on this file the output is as following:

```
Inspecting 1 file
C

Offenses:

/tmp/test.rb:1:6: C: Use next to skip iteration.
foos.each do
     ^^^^
/tmp/test.rb:2:3: C: Favor modifier unless usage when having a single-line body. Another good alternative is the usage of control flow &&/||.
  unless foo
  ^^^^^^

1 file inspected, 2 offenses detected
```

The second warning makes sense but the first one is absolute nonsense. That is, the following code is no better:

```ruby
foos.each do
  next if foo
  yield bar
end
```

To make things more confusing, the following *doesn't* produce the above warning:

```ruby
foos.each do
  yield bar unless foo
end
```

As a more real-world example, there's the following snippet of code that we have:

```ruby
review_nodes.each do |review|
  unless review.attr('class').include?('separator')
    yield review
  end
end
```

Rewriting this as following brings absolutely no benefits:

```ruby
review_nodes.each do |review|
  next if review.attr('class').include?('separator')

  yield review
end
```

It *might* save you a single level of indentation but that's about it. It *might* make sense if you have multiple conditionals that determine whether or not to skip the loop. However, in simple cases such as those shown above the warning is utter nonsense.

Another snippet, which also contains a confusing line about where the error occurs:

```ruby
ReviewIterator.new(page).each do |review|
  parsed = parse_review(review)

  unless review_exists?(config, parsed[:review_id])
    reviews << parsed
  end
end
```

This result in:

```
Inspecting 1 file
C

Offenses:

/tmp/test.rb:1:26: C: Use next to skip iteration.
ReviewIterator.new(page).each do |review|
                         ^^^^
/tmp/test.rb:4:3: C: Favor modifier unless usage when having a single-line body. Another good alternative is the usage of control flow &&/||.
  unless review_exists?(config, parsed[:review_id])
  ^^^^^^

1 file inspected, 2 offenses detected
```

Again rewriting this gives no benefits, unless more complex code were to be used as the conditional:

```
ReviewIterator.new(page).each do |review|
  parsed = parse_review(review)

  next if review_exists?(config, parsed[:review_id])

  reviews << parsed
end
```

I realize that it's rather difficult to figure out the complexity of code following a `next` and when to throw warnings or not, but throwing this warning *always* is a massive overkill.

**Comment by AlexCppns** [comment]
Is there a way to disable this cop?

**Comment by bbatsov** [comment]
Of course - you can disable it just like any other cop in `.rubocop.yml`.
